File: tools/ticket_issuance_planner_v2/plan.py
from .core import *  # noqa: F401,F403
from .drafts import load_draft_pair
from .context import validate_context, validate_registries, validate_control_schema
from .control import (
    selection_state,
    validate_root_metadata_and_package_state,
    validate_workflows,
    validate_handoffs,
    validate_output,
    choose_decision,
)


def build_plan(
    args: argparse.Namespace,
) -> tuple[dict[str, Any], Path | None]:
    root = Path(args.root).resolve()
    checks = Checks()
    state = selection_state(
        args.base_commit, args.writer, args.reviewer, checks
    )

    selected_for_view = args.base_commit if args.base_commit is not None else None
    try:
        view = GitView(root, selected_for_view)
        checks.pass_(
            "git-view",
            f"all repository inputs read from immutable commit {view.tagged_commit}",
        )
    except PlannerFailure as exc:
        if exc.reason_code != "BASE_COMMIT_INVALID":
            raise
        checks.fail("base-commit", exc.reason_code, exc.message)
        view = GitView(root, None)
        checks.pass_(
            "git-view-fallback",
            f"invalid selection inspected against immutable HEAD {view.tagged_commit}",
        )

    launch, package_row, function_row, stage_row = validate_registries(
        view, args.package, checks
    )
    pair = load_draft_pair(
        view, args.package, package_row, checks
    )
    sources = (
        validate_context(view, pair, args.package, checks) if pair else []
    )
    validate_control_schema(view, launch, checks)
    validate_root_metadata_and_package_state(view, args.package, checks)
    validate_workflows(view, checks)
    handoffs = (
        validate_handoffs(
            view, pair, args.accepted_handoff, checks
        )
        if pair
        else []
    )

    classification = "UNKNOWN"
    if (
        isinstance(launch.get("authorized_packages"), list)
        and launch["authorized_packages"].count(args.package) == 1
    ):
        classification = "AUTHORIZED"
    elif (
        isinstance(launch.get("conditional_packages"), list)
        and launch["conditional_packages"].count(args.package) == 1
    ):
        classification = "CONDITIONAL"
    if (
        pair
        and pair.ticket.get("launch_class") == classification
        and classification in {"AUTHORIZED", "CONDITIONAL"}
    ):
        checks.pass_(
            "launch-class",
            f"draft and launch classification agree: {classification}",
        )
    else:
        checks.fail(
            "launch-class",
            "PACKAGE_STAGE_MISMATCH",
            "draft and launch classification differ",
        )

    output_target = validate_output(root, args.output, checks)
    decision = choose_decision(state, checks.reasons)
    package_path = package_row.get("path", "") if package_row else ""
    package_wave = package_row.get("wave", -1) if package_row else -1
    scope = function_row.get("write_scope", "") if function_row else ""

    plan: dict[str, Any] = {
        "schema_version": SCHEMA_VERSION,
        "record_kind": RECORD_KIND,
        "status": STATUS,
        "repository": {
            "name": REPOSITORY_NAME,
            "view_commit": view.tagged_commit,
            "object_format": view.object_format,
            "working_tree_used_as_input": False,
        },
        "package": {
            "name": args.package,
            "path": package_path,
            "wave": package_wave,
            "write_scope": scope,
            "source_ceiling_class": (
                pair.source_ceiling_class if pair else ""
            ),
        },
        "stage": {
            "id": pair.ticket.get("stage", "UNKNOWN")
            if pair
            else "UNKNOWN",
            "phase": pair.ticket.get("phase", "UNKNOWN")
            if pair
            else "UNKNOWN",
            "active_stage": launch.get("active_stage", "UNKNOWN"),
            "active_wave": launch.get("active_wave", -1),
            "registry_wave": stage_row.get("wave", -1)
            if stage_row
            else -1,
        },
        "launch": {
            "classification": classification,
            "classification_recognized": classification
            in {"AUTHORIZED", "CONDITIONAL"},
            "conditional_requirements": launch.get(
                "conditional_activation", {}
            ).get(args.package, {})
            if isinstance(launch.get("conditional_activation"), dict)
            else {},
        },
        "selection": {
            "state": state,
            "base_commit": args.base_commit or "",
            "writer": args.writer or "",
            "reviewer": args.reviewer or "",
        },
        "drafts": {
            "ticket_path": pair.ticket_path if pair else "",
            "ticket_git_blob_id": pair.ticket_blob if pair else "",
            "ticket_exact_sha256": pair.ticket_sha256 if pair else "",
            "context_path": pair.context_path if pair else "",
            "context_git_blob_id": pair.context_blob if pair else "",
            "context_exact_sha256": pair.context_sha256 if pair else "",
            "sources": sources,
            "registry_selectors": list(pair.selectors) if pair else [],
            "accepted_handoff_slots": list(pair.handoff_slots)
            if pair
            else [],
            "unavailable_checks": list(pair.unavailable_checks)
            if pair
            else [],
        },
        "prerequisites": {"accepted_handoffs": handoffs},
        "checks": checks.items,
        "decision": decision,
        "reason_codes": checks.reasons,
        "mutations": [],
        "authorizes_context_materialization": False,
        "authorizes_ticket_issuance": False,
        "creates_writer_lease": False,
        "authorizes_implementation": False,
        "publishes_package_handoff": False,
        "advances_launch_state": False,
    }
    plan["plan_sha256"] = plan_digest(plan)
    return plan, output_target


def write_plan(plan: Mapping[str, Any], target: Path | None) -> None:
    encoded = canonical_json_bytes(plan)
    if len(encoded) > 262_144:
        raise PlannerFailure(
            "OUTPUT_WRITE_FAILED",
            "canonical plan exceeds 262144-byte ceiling",
        )
    if target is None:
        sys.stdout.buffer.write(encoded)
        return
    temporary = target.with_name(target.name + f".tmp-{os.getpid()}")
    try:
        target.parent.mkdir(parents=True, exist_ok=True)
        temporary.write_bytes(encoded)
        os.replace(temporary, target)
    except OSError as exc:
        try:
            temporary.unlink(missing_ok=True)
        except OSError:
            pass
        raise PlannerFailure(
            "OUTPUT_WRITE_FAILED",
            f"unable to write advisory artifact: {exc}",
        ) from exc

File: xtask/src/ticket_issuance_builder/assemble/plan.rs
//! Canonical advisory-plan projection.

use serde_json::{Value as JsonValue, json};
use toml::Value;

use crate::ticket_planner::{
    DECISION_INVALID, INVALID_REASONS, RECORD_KIND, REPOSITORY_NAME,
    SCHEMA_VERSION, STATUS, plan_digest,
};

use super::super::model::{
    Checks, DraftPair, RegistrySnapshot, TicketIssuanceBuildOptions,
};
use super::super::util::{count_string, integer, text, toml_to_json};

#[allow(clippy::too_many_arguments)]
pub(super) fn assemble_plan(
    options: &TicketIssuanceBuildOptions,
    tree: &crate::git_tree::GitTree,
    registries: &RegistrySnapshot,
    pair: Option<&DraftPair>,
    sources: Vec<JsonValue>,
    accepted_handoffs: Vec<JsonValue>,
    classification: &'static str,
    decision: &'static str,
    checks: &Checks,
) -> JsonValue {
    let package_path = registries
        .package_row
        .as_ref()
        .and_then(|row| text(row, "path"))
        .unwrap_or_default();
    let package_wave = registries
        .package_row
        .as_ref()
        .and_then(|row| integer(row, "wave"))
        .unwrap_or(-1);
    let scope = registries
        .function_row
        .as_ref()
        .and_then(|row| text(row, "write_scope"))
        .unwrap_or_default();
    let stage_id = pair
        .and_then(|pair| text(&pair.ticket, "stage"))
        .unwrap_or("UNKNOWN");
    let phase = pair
        .and_then(|pair| text(&pair.ticket, "phase"))
        .unwrap_or("UNKNOWN");
    let registry_wave = registries
        .stage_row
        .as_ref()
        .and_then(|row| integer(row, "wave"))
        .unwrap_or(-1);
    let conditional_requirements = registries
        .launch
        .get("conditional_activation")
        .and_then(Value::as_table)
        .and_then(|table| table.get(&options.package))
        .map_or_else(|| json!({}), toml_to_json);

    let mut plan = json!({
        "schema_version": SCHEMA_VERSION,
        "record_kind": RECORD_KIND,
        "status": STATUS,
        "repository": {
            "name": REPOSITORY_NAME,
            "view_commit": tree.tagged_commit(),
            "object_format": tree.object_format(),
            "working_tree_used_as_input": false,
        },
        "package": {
            "name": options.package.as_str(),
            "path": package_path,
            "wave": package_wave,
            "write_scope": scope,
            "source_ceiling_class": pair.map_or(
                "",
                |pair| pair.source_ceiling_class.as_str(),
            ),
        },
        "stage": {
            "id": stage_id,
            "phase": phase,
            "active_stage": text(&registries.launch, "active_stage")
                .unwrap_or("UNKNOWN"),
            "active_wave": integer(&registries.launch, "active_wave")
                .unwrap_or(-1),
            "registry_wave": registry_wave,
        },
        "launch": {
            "classification": classification,
            "classification_recognized": matches!(
                classification,
                "AUTHORIZED" | "CONDITIONAL"
            ),
            "conditional_requirements": conditional_requirements,
        },
        "selection": {
            "state": selection_state(options),
            "base_commit": options.base_commit.as_deref().unwrap_or(""),
            "writer": options.writer.as_deref().unwrap_or(""),
            "reviewer": options.reviewer.as_deref().unwrap_or(""),
        },
        "drafts": {
            "ticket_path": pair.map_or(
                "",
                |pair| pair.ticket_path.as_str(),
            ),
            "ticket_git_blob_id": pair.map_or(
                "",
                |pair| pair.ticket_blob.as_str(),
            ),
            "ticket_exact_sha256": pair.map_or(
                "",
                |pair| pair.ticket_sha256.as_str(),
            ),
            "context_path": pair.map_or(
                "",
                |pair| pair.context_path.as_str(),
            ),
            "context_git_blob_id": pair.map_or(
                "",
                |pair| pair.context_blob.as_str(),
            ),
            "context_exact_sha256": pair.map_or(
                "",
                |pair| pair.context_sha256.as_str(),
            ),
            "sources": sources,
            "registry_selectors": pair.map_or_else(
                Vec::new,
                |pair| pair.selectors.clone(),
            ),
            "accepted_handoff_slots": pair.map_or_else(
                Vec::new,
                |pair| pair.handoff_slots.clone(),
            ),
            "unavailable_checks": pair.map_or_else(
                Vec::new,
                |pair| pair.unavailable_checks.clone(),
            ),
        },
        "prerequisites": {
            "accepted_handoffs": accepted_handoffs,
        },
        "checks": checks.checks_json(),
        "decision": decision,
        "reason_codes": checks.reasons(),
        "mutations": [],
        "authorizes_context_materialization": false,
        "authorizes_ticket_issuance": false,
        "creates_writer_lease": false,
        "authorizes_implementation": false,
        "publishes_package_handoff": false,
        "advances_launch_state": false,
    });
    let digest = plan_digest(&plan);
    plan.as_object_mut()
        .expect("ticket plan is an object")
        .insert("plan_sha256".to_owned(), JsonValue::String(digest));
    debug_assert_eq!(
        decision == DECISION_INVALID,
        checks
            .reasons()
            .iter()
            .any(|reason| INVALID_REASONS.contains(&reason.as_str()))
    );
    plan
}

pub(super) fn launch_class(
    launch: &Value,
    package: &str,
) -> &'static str {
    if count_string(launch.get("authorized_packages"), package) == 1 {
        "AUTHORIZED"
    } else if count_string(launch.get("conditional_packages"), package) == 1 {
        "CONDITIONAL"
    } else {
        "UNKNOWN"
    }
}

fn selection_state(options: &TicketIssuanceBuildOptions) -> &'static str {
    let count = [
        options.base_commit.as_ref(),
        options.writer.as_ref(),
        options.reviewer.as_ref(),
    ]
    .iter()
    .filter(|value| value.is_some())
    .count();
    match count {
        0 => "NONE",
        3 => "COMPLETE",
        _ => "PARTIAL",
    }
}
